ddtrace/contrib/urllib3/patch.py

```diff
@@ -0,0 +1,131 @@
+    },
+)
+
+
+def patch():
+    """Enable tracing for all urllib3 requests"""
+    if getattr(urllib3, "__datadog_patch", False):
+        return
+    setattr(urllib3, "__datadog_patch", True)
+
+    _w("urllib3", "connectionpool.HTTPConnectionPool.urlopen", _wrap_urlopen)
+
+
+def unpatch():
+    """Disable trace for all urllib3 requests"""
+    if getattr(urllib3, "__datadog_patch", False):
+        setattr(urllib3, "__datadog_patch", False)
+
+        _u(urllib3.connectionpool.HTTPConnectionPool, "urlopen")
+
+
+def _wrap_urlopen(func, instance, args, kwargs):
+    """
+    Wrapper function for the lower-level urlopen in urllib3
+
+    :param func: The original target function "urlopen"
+    :param instance: The patched instance of ``HTTPConnectionPool``
+    :param args: Positional arguments from the target function
+    :param kwargs: Keyword arguments from the target function
+    :return: The ``HTTPResponse`` from the target function
+    """
+    request_method = get_argument_value(args, kwargs, 0, "method")
+    request_url = get_argument_value(args, kwargs, 1, "url")
+    try:
+        request_headers = get_argument_value(args, kwargs, 3, "headers")
+    except ArgumentError:
+        request_headers = None
+    try:
+        request_retries = get_argument_value(args, kwargs, 4, "retries")
+    except ArgumentError:
+        request_retries = None
+
+    # HTTPConnectionPool allows relative path requests; convert the request_url to an absolute url
+    if request_url.startswith("/"):
+        request_url = parse.urlunparse(
+            (
+                instance.scheme,
+                "{}:{}".format(instance.host, instance.port)
+                if instance.port and instance.port not in DROP_PORTS
+                else str(instance.host),
+                request_url,
+                None,
+                None,
+                None,
+            )
+        )
+
+    parsed_uri = parse.urlparse(request_url)
+    hostname = parsed_uri.netloc
+
+    pin = Pin.get_from(instance)
+    if not pin or not pin.enabled():
+        return func(*args, **kwargs)
+
+    with pin.tracer.trace(
+        "urllib3.request", service=trace_utils.ext_service(pin, config.urllib3), span_type=SpanTypes.HTTP
+    ) as span:
+        if config.urllib3.split_by_domain:
+            span.service = hostname
+
+        # If distributed tracing is enabled, propagate the tracing headers to downstream services
+        if config.urllib3.distributed_tracing:
+            if request_headers is None:
+                request_headers = {}
+                kwargs["headers"] = request_headers
+            HTTPPropagator.inject(span.context, request_headers)
+
+        if config.urllib3.analytics_enabled:
+            span.set_tag(ANALYTICS_SAMPLE_RATE_KEY, config.urllib3.get_analytics_sample_rate())
+        if isinstance(request_retries, urllib3.util.retry.Retry):
+            span.set_tag(http.RETRIES_REMAIN, str(request_retries.total))
```

Comment:
Should this be moved to `set_http_meta`?

Reply:
Makes sense, but I don't think we want to treat it as a metric?